[VPlan] Handle single-scalar conds in VPWidenSelectRecipe.

fhahn opened the pull request:
Generalize VPWidenSelectRecipe codegen to consider single-scalar conditions instead of just loop-invariant ones.

If the condition is a single-scalar, we can simply use a scalar condition.

fhahn committed:
[VPlan] Handle single-scalar conds in VPWidenSelectRecipe.

Generalize VPWidenSelectRecipe codegen to consider single-scalar
conditions instead of just loop-invariant ones.

If the condition is a single-scalar, we can simply use a scalar
condition.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1757,16 +1757,8 @@ struct LLVM_ABI_FOR_TEST VPWidenSelectRecipe : public VPRecipeWithIRFlags,
     return getOperand(0);
```
Worth documenting somewhere that widen select recipe follows the select instruction in supporting both a widen and a uniform condition.

fhahn replied:
Added, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-select-single-scalar-cond

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2022,6 +2023,11 @@ InstructionCost VPWidenSelectRecipe::computeCost(ElementCount VF,
 
+bool VPWidenSelectRecipe::onlyFirstLaneUsed(const VPValue *Op) const {
+  assert(is_contained(operands(), Op) && "Op must be an operand of the recipe");
+  return Op == getCond() && vputils::isSingleScalar(getCond());
```
Yes, but also needed to determine if we need to broadcast or not at the moment, for which we need to provide this at the uses as well, I think?

lukel97 replied:
I guess it's a bit strange but fine by me. Another idea could be to simplify a VPWidenSelectRecipe with a single-scalar condition to a VPInstruction select?

fhahn replied:
I updated the patch to implement this as simplification of broadcasts, together with https://github.com/llvm/llvm-project/pull/165826

ayalz commented on llvm/test/Transforms/LoopVectorize/AArch64/epilog-iv-select-cmp.ll:
```diff
@@ -48,7 +48,7 @@ define i8 @select_icmp_var_start(ptr %a, i8 %n, i8 %start) {
 ; CHECK-NEXT:    [[IND_END:%.*]] = trunc i32 [[N_VEC]] to i8
 ; CHECK-NEXT:    [[MIN_EPILOG_ITERS_CHECK:%.*]] = icmp ult i32 [[N_MOD_VF]], 8
-; CHECK-NEXT:    br i1 [[MIN_EPILOG_ITERS_CHECK]], label %[[VEC_EPILOG_SCALAR_PH]], label %[[VEC_EPILOG_PH]]
+; CHECK-NEXT:    br i1 [[MIN_EPILOG_ITERS_CHECK]], label %[[VEC_EPILOG_SCALAR_PH]], label %[[VEC_EPILOG_PH]], !prof [[PROF3:![0-9]+]]
```
How/is this change related?

fhahn committed:
[VPlan] Remove RepOrWidenR

fhahn committed:
Restore

fhahn committed:
[VPlan] Convert BuildVector with all-equal values to Broadcast.

Fold BuildVector where all operands are equal to Broadcast of the first
operand. This will subsequently make it easier to remove additional
buildvectors/broadcasts, e.g. via
https://github.com/llvm/llvm-project/pull/165506.

fhahn committed:
[VPlan] BuildVector to broadcast.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-select-single-scalar-cond

fhahn committed:
!fixup move to transform, support Select VPInsts

fhahn committed:
Merge branch 'vplan-buildvector-to-broadcast' into vplan-select-single-scalar-cond

lukel97 reviewed (approved):
LGTM

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-select-single-scalar-cond

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -659,7 +659,9 @@ Value *VPInstruction::generate(VPTransformState &State) {
   }
   case Instruction::Select: {
     bool OnlyFirstLaneUsed = vputils::onlyFirstLaneUsed(this);
```
The usage of OnlyFirstLaneUsed here and elsewhere deserves rethinking, its usage here essentially asks if the recipe deals with single scalar operands (and result), rather than using only first lanes if multiple lanes are provided.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1286,6 +1286,15 @@ static void simplifyRecipe(VPSingleDefRecipe *Def, VPTypeAnalysis &TypeInfo) {
+  // that case the scalar condition can be used directly.
+  if (match(Def,
+            m_Select(m_Broadcast(m_VPValue(C)), m_VPValue(), m_VPValue())) &&
+      vputils::isSingleScalar(C)) {
```
Is this redundant (can assert instead) - if Def is a Broadcast its operand C must be a single scalar?

ayalz reviewed (commented):
Couple of post-commit thoughts.
